include/circt/Dialect/FIRRTL/FIRRTLStatements.td

```diff
@@ -131,12 +131,16 @@ def FPrintFOp : FIRRTLOp<"fprintf"> {
   }];
 
   let arguments = (ins ClockType:$clock, UInt1Type:$cond, StrAttr:$outputFile,
+                       // Use FStringType instead of PrintfOperandType to reject hardware types.
+                       Variadic<FStringType>:$outputFileSubstitutions,
                        StrAttr:$formatString, Variadic<PrintfOperandType>:$substitutions,
                        StrAttr:$name);
   let results = (outs);
 
+  let hasVerifier = 1;
+
   let assemblyFormat = [{
-    $clock `,` $cond `,` $outputFile `,` $formatString ``
+    $clock `,` $cond `,` $outputFile (` ` `(` $outputFileSubstitutions^ `)`)? `,` $formatString ``
```

Comment:
Mega-nit: the `formatString`, when serialized, doesn't have a space after it, but the `outputFile` format string does. It at least shows this in the `parse-basic.fir` test. Is there a way to make this the same?

Reply:
It should be fixed in hand-written printer.